File: simple_python_package/simple_python_package/simple_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy, QoSHistoryPolicy

class PublisherNode(Node):
    def __init__(self):
        super().__init__('publisher_node')
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.RELIABLE, durability=QoSDurabilityPolicy.TRANSIENT_LOCAL,  history=QoSHistoryPolicy.KEEP_LAST, depth=10)
        
        self.publisher = self.create_publisher(String, 'chatter', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.i = 0

    def timer_callback(self):
        msg = String()
        msg.data = f'Hello, ROS2! Count: {self.i}'
        self.publisher.publish(msg)
        self.get_logger().info(f'Publishing: "{msg.data}"')
        self.i += 1

def main(args=None):
    rclpy.init(args=args)

    publisher_node = PublisherNode()

    rclpy.spin(publisher_node)

    publisher_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: simple_cpp_package/src/simple_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int32.hpp"
#include "chrono"

using namespace std::chrono_literals;


class SimplePublisher:public rclcpp::Node
{
public: 
        SimplePublisher(): Node("publisher_node"),count_(0)
        {
            //message = std::make_shared<std_msgs::msg::Int32>(); message->data = 0;

            publisher_ = this->create_publisher<std_msgs::msg::Int32>("counter_2", 10);

            timer_ = this -> create_wall_timer(500ms, std::bind(&SimplePublisher::timer_callback, this));
            // timer_ = this -> create_wall_timer(500ms, [this]()
            // {
            //     timer_callback();
            //     // std::cout << "Timer callback executed. count_ = " << message->data << std::endl;
            //     });
        }

        // void timer_callback()
        // {
        // if (message) 
        //     {
        //         message->data++;
        //         publisher_->publish(*message);
        //     } 
        // else
        //     {
        //         std::cout << "Error: message is null." << std::endl;
        //     }
        // }

        

        void timer_callback()
        {
            
            auto message = std_msgs::msg::Int32();
            message.data = count_;
            count_++;
            std::cout<<"msg.data: "<<message.data<<std::endl;
            publisher_->publish(message);
        }
        /*Destructor*/

        ~SimplePublisher()
        {
            std::cout << "Destructor called with value: "<< count_<< std::endl;
        }
private: 
        //std::shared_ptr<std_msgs::msg::Int32> message;
        size_t count_;
        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr publisher_;

};
int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimplePublisher>());
    rclcpp::shutdown();
    return 0;
}